common/tools/eslint-plugin-azure-sdk/src/rules/ts-doc-internal-private-member.ts

```diff
@@ -0,0 +1,154 @@
+ */
+const reportInternal = (
+  node: Node,
+  context: Rule.RuleContext,
+  converter: ParserWeakMapESTreeToTSNode,
+  typeChecker: TypeChecker
+): void => {
+  const tsNode = converter.get(node as TSESTree.Node) as any;
+  const symbol = typeChecker.getTypeAtLocation(tsNode).getSymbol();
+
+  const modifiers = converter.get(node as TSESTree.Node).modifiers;
+  // if type is internal and has a TSDoc and is a private member
+  if (!context.settings.exported.includes(symbol) && tsNode.jsDoc !== undefined
+      && modifiers !== undefined && modifiers.some(
+      (modifier: Modifier): boolean => modifier.kind === SyntaxKind.PrivateKeyword)
+  ) {
+    // fetch all tags
+    let TSDocTags: string[] = [];
+    tsNode.jsDoc.forEach((TSDocComment: any): void => {
+      TSDocTags = TSDocTags.concat(
+        TSDocComment.tags !== undefined
+          ? TSDocComment.tags.map((TSDocTag: any): string => TSDocTag.tagName.escapedText)
+          : []
+      );
+    });
+
+    // see if any tags match internal
+    if (TSDocTags.some((TSDocTag: string): boolean => /(internal)/.test(TSDocTag))) {
+      context.report({
+        node: node,
+        message: "private class members should not include an @internal tag"
+      });
+    }
+  }
+};
+
+/**
+ * Determine whether this rule should examine a given file
+ * @param fileName the filename of the file in question
+ * @param exclude the list of files excluded by TypeDoc (other than those in node_modues)
+ * @returns false if not in src or is excluded by TypeDoc
+ */
+const shouldExamineFile = (fileName: string, exclude: string[]): boolean => {
+  if (!/src/.test(fileName)) {
+    return false;
+  }
+  const relativePath = relative("", fileName).replace(/\\/g, "/");
+  return !exclude.includes(relativePath);
+};
+
+let exclude: string[] = [];
+try {
+  const typeDocText = readFileSync("typedoc.json", "utf8");
+  const typeDoc = JSON.parse(typeDocText);
+
+  // if typeDoc.exclude exists, add all files matching the glob patterns to exclude
+  if (typeDoc.exclude !== undefined) {
+    typeDoc.exclude.forEach((excludedGlob: string): void => {
+      exclude = exclude.concat(
+        globSync(excludedGlob).filter(
+          (excludeFile: string): boolean => !/node_modules/.test(excludeFile)
+        )
+      );
+    });
+  }
+} catch (err) {
+  exclude = [];
+}
+
+export = {
+  meta: getRuleMetaData(
+    "ts-doc-internal-private-member",
+    "requires TSDoc comments to not include an '@internal' tag if the object is private"
+  ),
+  create: (context: Rule.RuleContext): Rule.RuleListener => {
+    const fileName = context.getFilename();
+
+    // on the first run, if on a .ts file (program.getSourceFile is file-type dependent)
+    if (context.settings.exported === undefined && /\.ts$/.test(fileName)) {
+      const packageExports = getLocalExports(context);
+      if (packageExports !== undefined) {
+        context.settings.exported = packageExports;
```

Comment:
The problem is caused by this line. It somehow breaks the `import/no-extraneous-dependencies` rule https://github.com/import-js/eslint-plugin-import/blob/e8d79b5abe295eb956401b06d7dfb87b29ad5635/src/rules/no-extraneous-dependencies.js

I read about `context.settings` can be used to store data that are accessible to all rules. However, I didn't see it used explicitly in the failing rule.  Maybe this data is written to file then caused problem.

Our other rule `ts-doc-internal` also caused the same error, but it's currently disabled becaues of a different issue.
https://github.com/Azure/azure-sdk-for-js/blob/355941804f6899ae9a7a81f578dc575e8999436e/common/tools/eslint-plugin-azure-sdk/src/configs/azure-sdk-base.ts#L124-L125

Reply:
I guess one workaround is to use a local cache instead of `context.settings.exported`. We will need to see how much it would impact total linting time.

Reply:
@jeremymeng Thanks for investigating this!

Yeah I agree with local caching, here is an example: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/consistent-type-exports.ts

Reply:
Great, thank you so much for looking into this and pointing me to resources! I'll make the change to local caching and see if that fixes it

Reply:
> Yeah I agree with local caching, here is an example: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/consistent-type-exports.ts

This looks a lot more than what I was thinking.  My suggestion is just to replace `context.settings.exported` with a variable `let exportedCache: Symbol[]` at the file/module level. @deyaaeldeen any drawback you can think of?

Reply:
@jeremymeng the cache in the rule I shared is implemented at the create method level, see https://github.com/typescript-eslint/typescript-eslint/blob/8377e6ea422ee2d52455da8955ff055e09c238d3/packages/eslint-plugin/src/rules/consistent-type-exports.ts#L55. Is there a reason to have the cache at the file/module level?

Reply:
I see.  `context.settings.exported` is used in `reportInternal` so I was thinking about module level, but we could pass it as a parameter if it's declared in `create`

Reply:
I made the change to local caching but then realized that the exported settings weren't needed for this specific rule. I'm not sure why it was there in the `ts-doc-internal` rule but I removed it from my code and it works now

Reply:
You are right! `ts-doc-internal` checks for things that are not exported so it builds a cache of all exported things.  This rule only targets private members.